### PR TITLE
Piling Coordinates

### PR DESCRIPTION
**Pull request by MichaelLiddiard**
Basic piling implementation. Further work is needed before production ready, but is needed to form base of other branches.

### CODE REVIEW

File: Cedar.sln [hunk 1]
```diff
@@ -3,10 +3,12 @@ Microsoft Visual Studio Solution File, Format Version 12.00
...
-Project("{FAE04EC0-301F-11D3-BF4B-00C04F79EFBC}") = "Piling", "src\Piling\Piling.csproj", "{19C78C67-15E5-4768-9018-9730FD14F1E3}"
+Project("{9A19103F-16F7-4668-BE54-9A1E7A4F7556}") = "Piling", "src\Piling\Piling.csproj", "{19C78C67-15E5-4768-9018-9730FD14F1E3}"
 EndProject
 Project("{FAE04EC0-301F-11D3-BF4B-00C04F79EFBC}") = "Cedar", "src\Cedar\Cedar.csproj", "{784517B7-0F92-4401-BCBF-CD2E1FB5FDFB}"
 EndProject
+Project("{FAE04EC0-301F-11D3-BF4B-00C04F79EFBC}") = "CedarCore", "src\CedarCore\CedarCore.csproj", "{27EE896C-A09A-4A5D-A995-C2A4043FB871}"
+EndProject
 Global
 	GlobalSection(SolutionConfigurationPlatforms) = preSolution
 		Debug|Any CPU = Debug|Any CPU
```
Comment (jon-r-abbott) [inline Cedar.sln]: Platform naming to be reviewed

File: Cedar.sln [hunk 2]
```diff
@@ -21,6 +23,10 @@ Global
 		{784517B7-0F92-4401-BCBF-CD2E1FB5FDFB}.Debug|Any CPU.Build.0 = Debug|Any CPU
 		{784517B7-0F92-4401-BCBF-CD2E1FB5FDFB}.Release|Any CPU.ActiveCfg = Release|Any CPU
 		{784517B7-0F92-4401-BCBF-CD2E1FB5FDFB}.Release|Any CPU.Build.0 = Release|Any CPU
+		{27EE896C-A09A-4A5D-A995-C2A4043FB871}.Debug|Any CPU.ActiveCfg = Debug|Any CPU
+		{27EE896C-A09A-4A5D-A995-C2A4043FB871}.Debug|Any CPU.Build.0 = Debug|Any CPU
+		{27EE896C-A09A-4A5D-A995-C2A4043FB871}.Release|Any CPU.ActiveCfg = Release|Any CPU
+		{27EE896C-A09A-4A5D-A995-C2A4043FB871}.Release|Any CPU.Build.0 = Release|Any CPU
```
Comment (jon-r-abbott) [inline Cedar.sln]: Platform naming to be reviewed

File: src/Cedar/Cedar.csproj [hunk 3]
```diff
@@ -36,6 +37,7 @@
     <DefineConstants>TRACE</DefineConstants>
     <ErrorReport>prompt</ErrorReport>
     <WarningLevel>4</WarningLevel>
+    <PlatformTarget>x64</PlatformTarget>
   </PropertyGroup>
```
Comment (jon-r-abbott) [inline src/Cedar/Cedar.csproj:42]: Platform naming to be reviewed

File: src/Cedar/CedarApplication.cs [hunk 4]
```diff
@@ -1,40 +1,59 @@
...
 
+        /// <summary>
+        /// Initialize components once the application has loaded
+        /// </summary>
+        /// <param name="sender">Application</param>
+        /// <param name="e">Event args</param>
+        private void ControlledApplication_ApplicationInitialized(object sender, ApplicationInitializedEventArgs e)
+        {
+            //Initialize shared components
+            ISharedParameterManager sharedParameterManager = new SharedParameterManager(sender as Application);
```
Comment (jon-r-abbott) [inline src/Cedar/CedarApplication.cs]: Possible null reference. Either explicit cast with () operator, check for null value or pattern match with 'is'.

File: src/Cedar/CedarApplication.cs [hunk 5]
```diff
@@ -1,40 +1,59 @@
...
+
+            //Register updaters
+            PilingUpdater.Register(sender as Application, sharedParameterManager);
```
Comment (jon-r-abbott) [inline src/Cedar/CedarApplication.cs]: Possible null reference. Either explicit cast with () operator, check for null value or pattern match with 'is'.

File: src/CedarCore/CedarCore.csproj [hunk 6]
```diff
@@ -0,0 +1,89 @@
...
+    <PlatformTarget>x64</PlatformTarget>
+  </PropertyGroup>
+  <PropertyGroup Condition=" '$(Configuration)|$(Platform)' == 'Release|AnyCPU' ">
+    <DebugType>pdbonly</DebugType>
+    <Optimize>true</Optimize>
+    <OutputPath>bin\Release\</OutputPath>
+    <DefineConstants>TRACE</DefineConstants>
+    <ErrorReport>prompt</ErrorReport>
+    <WarningLevel>4</WarningLevel>
+    <PlatformTarget>x64</PlatformTarget>
```
Comment (jon-r-abbott) [inline src/CedarCore/CedarCore.csproj:38]: Platform naming to be reviewed

File: src/CedarCore/ISharedParameterManager.cs [hunk 7]
```diff
@@ -0,0 +1,16 @@
...
+using System.Threading.Tasks;
+
+namespace Jpp.Cedar.Core
+{
+    public interface ISharedParameterManager
+    {
+        void BindParameters(Document document, Definition parameter, BuiltInParameterGroup group = BuiltInParameterGroup.INVALID);
+
+        Definition RegisterParameter(string groupName, string parameterName, ParameterType parameterType, bool editable, string description);
+    }
```
Comment (jon-r-abbott) [inline src/CedarCore/ISharedParameterManager.cs:30]: XML Comments

File: src/CedarCore/ISharedParameterManager.cs [hunk 8]
```diff
@@ -0,0 +1,16 @@
...
+using System.Text;
+using System.Threading.Tasks;
+
+namespace Jpp.Cedar.Core
+{
+    public interface ISharedParameterManager
+    {
+        void BindParameters(Document document, Definition parameter, BuiltInParameterGroup group = BuiltInParameterGroup.INVALID);
+
+        Definition RegisterParameter(string groupName, string parameterName, ParameterType parameterType, bool editable, string description);
```
Comment (jon-r-abbott) [inline src/CedarCore/ISharedParameterManager.cs:29]: Refactor method signature - possibly introduce class "SharedParameter" and use that.

File: src/CedarCore/Properties/AssemblyInfo.cs [hunk 9]
```diff
@@ -0,0 +1,36 @@
...
+// set of attributes. Change these attribute values to modify the information
+// associated with an assembly.
+[assembly: AssemblyTitle("CedarCore")]
+[assembly: AssemblyDescription("")]
+[assembly: AssemblyConfiguration("")]
+[assembly: AssemblyCompany("")]
+[assembly: AssemblyProduct("CedarCore")]
+[assembly: AssemblyCopyright("Copyright ©  2019")]
+[assembly: AssemblyTrademark("")]
+[assembly: AssemblyCulture("")]
```
Comment (jon-r-abbott) [inline src/CedarCore/Properties/AssemblyInfo.cs:15]: Needs to be reviewed later

File: src/CedarCore/SharedParameterManager.cs [hunk 10]
```diff
@@ -0,0 +1,92 @@
...
+                //var test = RegisterParameter("Piling", "Easting", ParameterType.Length, false, "Easting");
+                bool instanceBindOK = bindingMap.Insert(parameter, instanceBinding);
+            }
+            catch (Exception e)
+            {
+                Debug.WriteLine(e.Message);
+            }
+        }
+
+        public Definition RegisterParameter(string groupName, string parameterName, ParameterType parameterType, bool editable, string description)
```
Comment (jon-r-abbott) [inline src/CedarCore/SharedParameterManager.cs:52]: Single responsibility - feel like the method is doing several things.

File: src/Cedar/CedarApplication.cs (hunk 5 above)
Comment (jon-r-abbott) [inline src/Cedar/CedarApplication.cs]: Repeated cast on sender.

File: src/Cedar/CedarApplication.cs [hunk 11]
```diff
@@ -1,40 +1,59 @@
...
+        /// Entry point for application
+        /// </summary>
+        /// <param name="application">Application being launched</param>
+        /// <returns>Result of add-in load</returns>
         public Result OnStartup(UIControlledApplication application)
         {
-            // Add a new ribbon panel
-            RibbonPanel ribbonPanel = application.CreateRibbonPanel("NewRibbonPanel");
+            if(application == null)
+                throw new System.ArgumentNullException(nameof(application));
```
Comment (jon-r-abbott) [inline src/Cedar/CedarApplication.cs:20]: What's the impact on revit when throwing an exception here? Could we not just return Result.Failed?

File: src/CedarCore/CoordinateHelper.cs [hunk 12]
```diff
@@ -0,0 +1,40 @@
...
+                //XYZ loc = (basePoint.Location as LocationPoint).Point;
+            }*/
+
+            ProjectPosition projectPosition = document.ActiveProjectLocation.get_ProjectPosition(XYZ.Zero);
+            // Create a translation vector for the offsets
+            XYZ translationVector = new XYZ(projectPosition.EastWest, projectPosition.NorthSouth, projectPosition.Elevation);
+            Transform translationTransform = Transform.CreateTranslation(translationVector);            
+            // Create a rotation for the angle about true north
+            //const double angleRatio = Math.PI / 180; // angle conversion factor
+            Transform rotationTransform = Transform.CreateRotationAtPoint(XYZ.BasisZ, projectPosition.Angle, XYZ.Zero);
```
Comment (jon-r-abbott) [inline src/CedarCore/CoordinateHelper.cs:33]: Disposable - consider disposing before losing scope

File: src/CedarCore/CoordinateHelper.cs [hunk 13]
```diff
@@ -0,0 +1,40 @@
...
+                double projectSurvpntX = basePoint.get_Parameter(BuiltInParameter.BASEPOINT_EASTWEST_PARAM).AsDouble();
+                double projectSurvpntY = basePoint.get_Parameter(BuiltInParameter.BASEPOINT_NORTHSOUTH_PARAM).AsDouble();
+                double projectSurvpntZ = basePoint.get_Parameter(BuiltInParameter.BASEPOINT_ELEVATION_PARAM).AsDouble();
+                //XYZ loc = (basePoint.Location as LocationPoint).Point;
+            }*/
+
+            ProjectPosition projectPosition = document.ActiveProjectLocation.get_ProjectPosition(XYZ.Zero);
+            // Create a translation vector for the offsets
+            XYZ translationVector = new XYZ(projectPosition.EastWest, projectPosition.NorthSouth, projectPosition.Elevation);
+            Transform translationTransform = Transform.CreateTranslation(translationVector);
```
Comment (jon-r-abbott) [inline src/CedarCore/CoordinateHelper.cs:30]: Disposable - consider disposing before losing scope

File: src/CedarCore/SharedParameterManager.cs [hunk 14]
```diff
@@ -0,0 +1,92 @@
...
+                if (pilingGroup == null)
+                {
+                    pilingGroup = defFile.Groups.Create(groupName);
+                }
+
+                result = pilingGroup.Definitions.get_Item(parameterName);
+
+                if (result == null)
+                {
+                    ExternalDefinitionCreationOptions newDefinition =
```
Comment (jon-r-abbott) [inline src/CedarCore/SharedParameterManager.cs:81]: Disposable - consider disposing before losing scope

File: src/Piling/PilingUpdater.cs [hunk 15]
```diff
@@ -0,0 +1,119 @@
...
+
+        public PilingUpdater(AddInId id, ISharedParameterManager spManager)
+        {
+            if (id == null)
+                throw new System.ArgumentNullException(nameof(id));
+
+            if (spManager == null)
+                throw new System.ArgumentNullException(nameof(spManager));
+
+            _appId = id;
```
Comment (jon-r-abbott) [inline src/Piling/PilingUpdater.cs]: Consider single line join and check.

_appId = id ?? throw new System.ArgumentNullException(nameof(id));

File: src/Piling/PilingUpdater.cs [hunk 16]
```diff
@@ -0,0 +1,119 @@
...
+            _updaterId = new UpdaterId(_appId, new Guid("ddb23f37-892e-4b43-9e8a-0ad8ff381b2b"));
+            _spManager = spManager;
```
Comment (jon-r-abbott) [inline src/Piling/PilingUpdater.cs]: Consider single line join and heck.

_spManager = spManager ?? throw new System.ArgumentNullException(nameof(spManager));

File: src/Piling/PilingUpdater.cs (hunk 15 above)
Comment (jon-r-abbott) [reply src/Piling/PilingUpdater.cs]: Feels like _appId is just local to the constructor?

File: src/Piling/PilingUpdater.cs [hunk 17]
```diff
@@ -0,0 +1,119 @@
...
+            if (application == null)
+                throw new System.ArgumentNullException(nameof(application));
+
+            if (sharedParameterManager == null)
+                throw new System.ArgumentNullException(nameof(sharedParameterManager));
+
+            PilingUpdater updater = new PilingUpdater(application.ActiveAddInId, sharedParameterManager);
+            UpdaterRegistry.RegisterUpdater(updater);
+            
+            ElementCategoryFilter filter = new ElementCategoryFilter(BuiltInCategory.OST_StructuralFoundation);
```
Comment (jon-r-abbott) [inline src/Piling/PilingUpdater.cs:42]: Disposable - consider disposing before losing scope

File: src/Piling/PilingUpdater.cs [hunk 18]
```diff
@@ -0,0 +1,119 @@
...
+            List<ElementId> modifiedElementIds = new List<ElementId>();
+            modifiedElementIds.AddRange(data.GetAddedElementIds());
+            modifiedElementIds.AddRange(data.GetModifiedElementIds());
+
+            foreach (ElementId id in modifiedElementIds)
+            {
+                Element foundation = document.GetElement(id);
+                if (foundation.Location != null)
+                {
+                    XYZ location = CoordinateHelper.GetWorldCoordinates(document, (foundation.Location as LocationPoint).Point);
```
Comment (jon-r-abbott) [inline src/Piling/PilingUpdater.cs]: Possible null reference on foundation.Location

File: src/CedarCore/CoordinateHelper.cs [hunk 19]
```diff
@@ -0,0 +1,40 @@
...
+            {
+                BasePoint basePoint = locationPoint as BasePoint;
+                Location svLoc = basePoint.Location;
+                double projectSurvpntX = basePoint.get_Parameter(BuiltInParameter.BASEPOINT_EASTWEST_PARAM).AsDouble();
+                double projectSurvpntY = basePoint.get_Parameter(BuiltInParameter.BASEPOINT_NORTHSOUTH_PARAM).AsDouble();
+                double projectSurvpntZ = basePoint.get_Parameter(BuiltInParameter.BASEPOINT_ELEVATION_PARAM).AsDouble();
+                //XYZ loc = (basePoint.Location as LocationPoint).Point;
+            }*/
+
+            ProjectPosition projectPosition = document.ActiveProjectLocation.get_ProjectPosition(XYZ.Zero);
```
Comment (jon-r-abbott) [inline src/CedarCore/CoordinateHelper.cs:27]: Don't know if the document will be required in future, if not then I'd consider passing in projectPosition.

File: src/CedarCore/ISharedParameterManager.cs (hunk 8 above)
Comment (MichaelLiddiard) [reply src/CedarCore/ISharedParameterManager.cs:29]: There is a built in class for this, but believe it hooks into native so dont want to use. Also dont want to duplicate a revit class to save a few parameters

File: src/CedarCore/SharedParameterManager.cs (hunk 10 above)
Comment (MichaelLiddiard) [reply src/CedarCore/SharedParameterManager.cs:52]: Either creates or gets the parameter if it exists

File: src/Cedar/CedarApplication.cs (hunk 11 above)
Comment (MichaelLiddiard) [reply src/Cedar/CedarApplication.cs:20]: Not sure. Worth investigating, but seem to implcitly fail

File: src/CedarCore/CoordinateHelper.cs (hunk 19 above)
Comment (MichaelLiddiard) [reply src/CedarCore/CoordinateHelper.cs]: Will be required to get the 3(!) other coordinate systems

Comment (MichaelLiddiard) [reply src/CedarCore/CoordinateHelper.cs:27]: Will be required to ge the 3(!) other coordinate systems

**Comment by MichaelLiddiard** [comment]
All disposables will be implemented with new base version in core when pushed.

File: src/CedarCore/SharedParameterManager.cs (hunk 10 above)
Comment (jon-r-abbott) [reply src/CedarCore/SharedParameterManager.cs:52]: Sorry, bit of a rushed comment. It wasn’t the get/create that concerned me. It was the creation of the file and setter it for the application, being in there too.

**Review by jon-r-abbott — APPROVED** [review]

**Pull request merged**